pkg/gce-pd-csi-driver/controller_test.go

```diff
@@ -1716,3 +1718,145 @@ func TestCreateVolumeDiskReady(t *testing.T) {
 		})
 	}
 }
+
+func TestControllerPublishVolume(t *testing.T) {
+	testCases := []struct {
+		name            string
+		seedDisks       []*gce.CloudDisk
+		req             *csi.ControllerPublishVolumeRequest
+		errorSeenOnNode bool
+	}{
+		{
+			name: "queue up requests if node has publish error",
+			seedDisks: []*gce.CloudDisk{
+				createZonalCloudDisk(name),
+			},
+			req: &csi.ControllerPublishVolumeRequest{
+				VolumeId: testVolumeID,
+				NodeId:   testNodeID,
+				VolumeCapability: &csi.VolumeCapability{
+					AccessType: &csi.VolumeCapability_Mount{
+						Mount: &csi.VolumeCapability_MountVolume{},
+					},
+					AccessMode: &csi.VolumeCapability_AccessMode{
+						Mode: csi.VolumeCapability_AccessMode_SINGLE_NODE_WRITER,
+					},
+				},
+			},
+			errorSeenOnNode: true,
+		},
+		{
+			name: "do not queue up requests if node doesn't have publish error",
+			seedDisks: []*gce.CloudDisk{
+				createZonalCloudDisk(name),
+			},
+			req: &csi.ControllerPublishVolumeRequest{
+				VolumeId: testVolumeID,
+				NodeId:   testNodeID,
+				VolumeCapability: &csi.VolumeCapability{
+					AccessType: &csi.VolumeCapability_Mount{
+						Mount: &csi.VolumeCapability_MountVolume{},
+					},
+					AccessMode: &csi.VolumeCapability_AccessMode{
+						Mode: csi.VolumeCapability_AccessMode_SINGLE_NODE_WRITER,
+					},
+				},
+			},
+			errorSeenOnNode: false,
+		},
+	}
+	for _, tc := range testCases {
+		t.Logf("test case: %s", tc.name)
+		gceDriver := initGCEDriver(t, tc.seedDisks)
+
+		// mark the node in the map
+		if tc.errorSeenOnNode {
+			gceDriver.cs.publishErrorsSeenOnNode[testNodeID] = true
+		}
+
+		requestCount := 50
+		for i := 0; i < requestCount; i++ {
+			gceDriver.cs.ControllerPublishVolume(context.Background(), tc.req)
+		}
+
+		if tc.errorSeenOnNode {
+			if err := wait.Poll(10*time.Nanosecond, 10*time.Millisecond, func() (bool, error) {
+				t.Logf("%v items queueed", gceDriver.cs.queue.Len())
+				return gceDriver.cs.queue.Len() == requestCount, nil
+			}); err != nil {
+				t.Fatalf("Only %v requests queued up for node has seen error", gceDriver.cs.queue.Len())
+			}
+		}
+
```

Comment:
If we wait a little longer, shouldn't the queue empty (because the fake provider will succeed the test?)

It's also possible that this test will be flaky. It would be more rigorous to use a fake rate limiter that could be controlled explicitly. I don't know if that's straightforward to do, so maybe we can see how this goes (although you may be creating more work for yourself in the long run...)

Reply:
The queue will be not empty as the request doesn't succeed so it will be re-queued. The reason why the request doesn't succeed is because I am using a real cloud provider so it will fail at getting volume lock step. The test case is to make sure when the request doesn't go through, all the requests failed and new requests will be queued. 

Maybe I can add more test cases using a fake cloud provider to have the request succeed but if that is the case, it is quite dynamic and async so nothing can be for sure.

Reply:
Added more tests